From e6e368f151c2461da94f7c2cd96bd80233f9175a Mon Sep 17 00:00:00 2001
From: "Wood, Brian J" <brian.j.wood@intel.com>
Date: Fri, 15 Mar 2019 11:37:47 -0700
Subject: [PATCH 2302/2304] media: pci: improper use of pointer before verified
 as valid

This fixes multiple uses of pointers before they've been
verified as having non-NULL value.


Change-Id: Ic9c4857d820e5a7362af97de290776df6f9335c2
Signed-off-by: Wood, Brian J <brian.j.wood@intel.com>
---
 drivers/media/pci/intel/ipu-wrapper.c | 78 +++++++++++++++++++++++------------
 1 file changed, 51 insertions(+), 27 deletions(-)

diff --git a/drivers/media/pci/intel/ipu-wrapper.c b/drivers/media/pci/intel/ipu-wrapper.c
index 9e06887..04ee03b 100644
--- a/drivers/media/pci/intel/ipu-wrapper.c
+++ b/drivers/media/pci/intel/ipu-wrapper.c
@@ -223,7 +223,8 @@ u64 shared_memory_alloc(unsigned int mmid, size_t bytes)
 	struct my_css_memory_buffer_item *buf;
 	unsigned long flags;
 
-	dev_dbg(mine->dev, "%s: in, size: %zu\n", __func__, bytes);
+	if (mine)
+		dev_dbg(mine->dev, "%s: in, size: %zu\n", __func__, bytes);
 
 	if (!bytes)
 		return (unsigned long)&alloc_cookie;
@@ -266,6 +267,11 @@ void shared_memory_free(unsigned int mmid, u64 addr)
 	struct my_css_memory_buffer_item *buf = NULL;
 	unsigned long flags;
 
+	if (!mine) {
+                pr_err("Invalid mem subsystem, return. mmid=%d", mmid);
+                return;
+        }
+
 	if ((void *)(unsigned long)addr == &alloc_cookie)
 		return;
 
@@ -306,6 +312,11 @@ u32 shared_memory_map(unsigned int ssid, unsigned int mmid, u64 addr)
 	struct my_css_memory_buffer_item *buf = NULL;
 	unsigned long flags;
 
+	if (!mine) {
+		pr_err("Invalid mem subsystem, return NULL. mmid=%d", mmid);
+		return NULL;
+	}
+
 	if ((void *)(unsigned long)addr == &alloc_cookie)
 		return 0;
 
@@ -337,9 +348,10 @@ void shared_memory_unmap(unsigned int ssid, unsigned int mmid, u32 addr)
  */
 void shared_memory_store_8(unsigned int mmid, u64 addr, u8 data)
 {
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%llx data = 0x%x\n",
-		__func__, addr, data);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%llx data = 0x%x\n",
+			__func__, addr, data);
 
 	*((u8 *)(unsigned long) addr) = data;
 	/*Invalidate the cache lines to flush the content to ddr. */
@@ -352,9 +364,10 @@ void shared_memory_store_8(unsigned int mmid, u64 addr, u8 data)
  */
 void shared_memory_store_16(unsigned int mmid, u64 addr, u16 data)
 {
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%llx data = 0x%x\n",
-		__func__, addr, data);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%llx data = 0x%x\n",
+			__func__, addr, data);
 
 	*((u16 *)(unsigned long) addr) = data;
 	/*Invalidate the cache lines to flush the content to ddr. */
@@ -367,9 +380,10 @@ void shared_memory_store_16(unsigned int mmid, u64 addr, u16 data)
  */
 void shared_memory_store_32(unsigned int mmid, u64 addr, u32 data)
 {
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%llx data = 0x%x\n",
-		__func__, addr, data);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%llx data = 0x%x\n",
+			__func__, addr, data);
 
 	*((u32 *)(unsigned long) addr) = data;
 	/* Invalidate the cache lines to flush the content to ddr. */
@@ -383,11 +397,12 @@ void shared_memory_store_32(unsigned int mmid, u64 addr, u32 data)
 void shared_memory_store(unsigned int mmid, u64 addr, const void *data,
 			 size_t bytes)
 {
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%lx bytes = 0x%zx\n", __func__,
-		(unsigned long)addr, bytes);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%lx bytes = 0x%zx\n", __func__,
+			(unsigned long)addr, bytes);
 
-	if (!data) {
+	if (!data && get_mem_sub_system(mmid)) {
 		dev_err(get_mem_sub_system(mmid)->dev,
 			"%s: data ptr is null\n", __func__);
 	} else {
@@ -409,9 +424,10 @@ void shared_memory_store(unsigned int mmid, u64 addr, const void *data,
  */
 void shared_memory_zero(unsigned int mmid, u64 addr, size_t bytes)
 {
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%llx data = 0x%zx\n",
-		__func__, (unsigned long long)addr, bytes);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%llx data = 0x%zx\n",
+			__func__, (unsigned long long)addr, bytes);
 
 	memset((void *)(unsigned long)addr, 0, bytes);
 	clflush_cache_range((void *)(unsigned long)addr, bytes);
@@ -425,8 +441,9 @@ u8 shared_memory_load_8(unsigned int mmid, u64 addr)
 {
 	u8 data = 0;
 
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%llx\n", __func__, addr);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%llx\n", __func__, addr);
 
 	/* Invalidate the cache lines to flush the content to ddr. */
 	clflush_cache_range((void *)(unsigned long)addr, sizeof(u8));
@@ -442,8 +459,9 @@ u16 shared_memory_load_16(unsigned int mmid, u64 addr)
 {
 	u16 data = 0;
 
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%llx\n", __func__, addr);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%llx\n", __func__, addr);
 
 	/* Invalidate the cache lines to flush the content to ddr. */
 	clflush_cache_range((void *)(unsigned long)addr, sizeof(u16));
@@ -459,8 +477,9 @@ u32 shared_memory_load_32(unsigned int mmid, u64 addr)
 {
 	u32 data = 0;
 
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%llx\n", __func__, addr);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%llx\n", __func__, addr);
 
 	/* Invalidate the cache lines to flush the content to ddr. */
 	clflush_cache_range((void *)(unsigned long)addr, sizeof(u32));
@@ -474,11 +493,12 @@ u32 shared_memory_load_32(unsigned int mmid, u64 addr)
  */
 void shared_memory_load(unsigned int mmid, u64 addr, void *data, size_t bytes)
 {
-	dev_dbg(get_mem_sub_system(mmid)->dev,
-		"access: %s: Enter addr = 0x%lx bytes = 0x%zx\n", __func__,
-		(unsigned long)addr, bytes);
+	if (get_mem_sub_system(mmid))
+		dev_dbg(get_mem_sub_system(mmid)->dev,
+			"access: %s: Enter addr = 0x%lx bytes = 0x%zx\n", __func__,
+			(unsigned long)addr, bytes);
 
-	if (!data) {
+	if (!data && get_mem_sub_system(mmid)) {
 		dev_err(get_mem_sub_system(mmid)->dev,
 			"%s: data ptr is null\n", __func__);
 
@@ -508,6 +528,10 @@ void ipu_wrapper_init(int mmid, struct device *dev, void __iomem *base)
 {
 	struct wrapper_base *sys = get_mem_sub_system(mmid);
 
+	if (!sys) {
+                pr_err("Invalid mem subsystem, return. mmid=%d", mmid);
+                return;
+        }
 	init_wrapper(sys);
 	sys->dev = dev;
 	sys->sys_base = base;
-- 
2.7.4

